library/fsio.c

```diff
@@ -0,0 +1,386 @@
+/*
+ *  Serialization based implementation of filesystem IO functions.
+ *
+ *  Copyright (C) 2017, ARM Limited, All Rights Reserved
+ *  SPDX-License-Identifier: Apache-2.0
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License"); you may
+ *  not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ *  WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ *  This file is part of mbed TLS (https://tls.mbed.org)
+ */
+
+#if !defined(MBEDTLS_CONFIG_FILE)
+#include "mbedtls/config.h"
+#else
+#include MBEDTLS_CONFIG_FILE
+#endif
+
+#if defined(MBEDTLS_FS_IO)
+
+#include <string.h>
+#include <assert.h>
+#include "mbedtls/fsio.h"
+
+#if !defined(MBEDTLS_FS_IO_ALT)
+
+#if !defined(MBEDTLS_PLATFORM_NO_STD_FUNCTIONS)
+/**
+ * \brief          Open file and disable buffering.
+ *
+ * \param path     File path
+ * \param mode     Open mode
+ *
+ * \return         File handle of type mbedtls_file_t.
+ *                 On failure MBEDTLS_FILE_INVALID is returned.
+ */
+mbedtls_file_t mbedtls_fopen( const char *path, const char *mode )
+{
+    mbedtls_file_t f = fopen( path, mode );
+
+    if( f != MBEDTLS_FILE_INVALID && ( setvbuf( f, NULL, _IONBF, 0 ) != 0 ) )
+    {
+        fclose( f );
+        f = MBEDTLS_FILE_INVALID;
+    }
+
+    return( f );
+}
+
+/**
+ * \brief           Read next directory entry.
+ *
+ * \note            This is rework of readdir to adapt to abstraction. This does
+ *                  not return struct dirent pointer. Hence no need to worry
+ *                  pointer ownership and cleanup.
+ *
+ * \param dir       Dir handle of type mbedtls_dir_t.
+ * \param file_name Out buffer for directory entry name.
+ *                  Up to 255 character long name can be returned.
+ * \param size      Out buffer length.
+ * \param type      Out pointer entry type.
+ *
+ * \return          0 for success. -1 for failure.
+ */
+int mbedtls_readdir( mbedtls_dir_t dir, char *file_name, uint32_t size,  uint32_t *type )
+{
+    int status = -1;
+    struct dirent *entry;
+
+    if ( ( entry = readdir( dir ) ) != NULL )
+    {
+        strncpy( file_name, entry->d_name, size );
+        if ( file_name[size - 1] == '\0' ) /* Check if buffer was enough */
+        {
+            *type = entry->d_type;
```

Comment:
If `d_type == DT_UNKNOWN`, you need to call `stat` to obtain the file type. Arguably, we shouldn't use `d_type` at all because it isn't portable: Glibc and *BSD have it but other POSIX systems may lack it. If there's no reliable way to detect the presence of `d_type` at compile time then we shouldn't assume that it's there. Efficiency is not critical here, we can just call `stat`.

Reply:
I can completely remove use of ```d_type```. However, ```stat``` would need full path of the file. Here ```readdir``` only returns the file name stripped of the the path. The user has to prefix dir path in order to use the file. 
Hence, for using ```stat``` either I have implement a separate API and user has to call that in addition to ```readdir``` or dir prefix has to be supplied in ```readdir``` call, which would look odd in presence of dir handle. 
What do you suggest? If it is not such a big issue I like the convenience offered by ```d_type```.

Reply:
`d_type` is convenient, sure, but it isn't always available. The [Linux man page](http://manpages.ubuntu.com/manpages/xenial/man2/getdents.2.html) warns that the availability depends on the filesystem type. I can imagine the support difficulties: “your library can't read my file! — Ok, send me the file. — Here you are. — Works for me. — Doesn't work for me! — Try a debugger... — Huh, `mbedtls_readdir` says the file doesn't exist.” Could take a while before somebody thinks of checking the filesystem type.

We could change `mbedtls_readdir` to not indicate the file type and provide a separate function for that. Alternatively, we could store the directory path in `mbedtls_dir_t`. Alternatively, there's `fstatat`, which exists on *BSD and Linux. It isn't POSIX though, so we should have a fallback for platforms such as POSIX compatibility layers on Windows, QNX, etc.

Reply:
I agree with you. I just wanted put different options to you. I Thought about storing the dir path when doing ```opendir```. But I am not comfortable storing the user supplied pointer to path. As it may change or re-used by the user. Other option is to copy but I neither want to allocate any memory. We can store it in the frontend. But that is only for target.
I like the idea of having a separate ```mbedtls_stat```. That we can limit to the fields we are interested in.